File: DragonIsland-GP2X/Background.cpp
#include "Background.h"

/**
 * Constructs the Background
 * @param fileNumber The number of the image file for the near and far layer images [0]-near, [1]-middle [2]-far
 * @param width The width of the level background
 * @param height The height of the level background
 * @param layerSpeed The scrolling speed of each background layer
 * @param alignement The alignment of the background 0-horizontal 1-vertical
 */
Background::Background(int fileNumber[], int width, int height, int layerSpeed[], int alignement)
{
	mFileNumber = fileNumber;
	mLayerX = new int[3];
	mWidth = width;
	mHeight = height;
	mAlign = alignement;
	mLayerSpeed = layerSpeed;
	mLayerImage = new Image[3];
	
	//load background layer images
	char path[100];
	sprintf (path, "mnt/sd/Dragon Island/bgr/near/%d.png", mFileNumber[0]);	
	mLayerImage[0] = *new Image(path);   //near
	sprintf (path, "mnt/sd/Dragon Island/bgr/middle/%d.png", mFileNumber[1]);	
	mLayerImage[1] = *new Image(path);  //middle
	sprintf (path, "mnt/sd/Dragon Island/bgr/far/%d.png", mFileNumber[2]);
	mLayerImage[2] = *new Image(path);  //far
	
	//set color from far layer top left pixel
	mColor = mLayerImage[2].get_pixel32(0, 0);
	
	//set the height using the LayerOffsetY values
	setHeight(height);
}

/**
 * Reads RGB colour and y offset value for the background layer
 * @param layer The layer name near, middle or far
 * @param fileNumber The file number of the background layer
 * @return The RGB color and y offset values
 */
vector<string> Background::readIni(char* layer, int fileNumber) {
	vector<string> tokens;
	char path[100];
	sprintf(path, "mnt/sd/Dragon Island/bgr/%s/%d.ini", layer, fileNumber);	
	ifstream file(path); // declare file stream: http://www.cplusplus.com/reference/iostream/ifstream/
	string line;
	while (file.good())
	{
		getline(file, line, ','); // read a string until next comma: http://www.cplusplus.com/reference/string/getline/
		tokens.push_back(line);		    
	}
	file.close();
    return tokens; 	
}

/**
 * Sets the speed of a background layer
 * @param layer The background layer 0-near, 1-middle 2-far
 * @param speed The horizontal scroll speed
 */
void Background::setLayerSpeed(int layer, int speed) {
    mLayerSpeed[layer] = speed;
}

/**
 * Sets the background color
 * @param colour The RGB background color
 */
void Background::setColor(Color* colour) {
    mColor = colour;
}

/**
 * Returns the background color
 * @return The RGB background color
 */
Color* Background::getColor() {
    return mColor;
}

/**
 * Sets the width of the background
 * @param width The width of the background
 */
void Background::setWidth(int width) {
    mWidth = width;
}

/**
 * Sets the height of the background
 * @param height The height of the background
 */
void Background::setHeight(int height) {
    mLayerY = new int[3];
    mLayerY[0] = height - mLayerImage[0].getHeight();
    mLayerY[1] = height - mLayerImage[1].getHeight();
    mLayerY[2] = height - mLayerImage[2].getHeight();
}

/**
 * Sets the alignment of the background
 * @param alignment the alignment of the background 0-repeat x axis / 1=repeat y axis (far layer)
 */
void Background::setAlign(int alignment) {
    mAlign = alignment;
}

/**
 * Draws the background on the graphics surface at the camera position
 * @param g The drawing surface
 * @param cam The camera coordinates
 * @return The camera coordinates within the screen area
 */
Point* Background::draw(SDL_Surface* g, Point* cam) {

	SDL_Rect rect;
	rect.x = 0;
	rect.y = 0;
	rect.w = 480;  
	rect.h = 272;
	SDL_FillRect(g, &rect, SDL_MapRGB(g->format , mColor->getRed(), mColor->getGreen(), mColor->getBlue()));

    if (cam->x >= mWidth - 480) {
        cam->x = mWidth - 480;
    }
    if (cam->x <= 0) {
        cam->x = 0;
    }
    if (cam->y >= mHeight - 272) {
        cam->y = mHeight - 272;
    }
    if (cam->y <= 0) {
        cam->y = 0;
    }

    int bgX = cam->x - (cam->x / 512 * 512);
    int bgY = cam->y - (cam->y / 512 * 512);

    if (Settings::Animation) //move background
    {
        if (mScrollTimer < 7) {
            mScrollTimer++;
        } else {
            for (int i = 0; i < 3; i++) {
                if (mLayerX[i] < 512) {
                    mLayerX[i] += mLayerSpeed[i];
                } else {
                    mLayerX[i] = 0;
                }
            }
            mScrollTimer = 0;
        }
    } else {
        mLayerX[0] = 0;
        mLayerX[1] = 0;
    }


    if (Settings::Background) {
        if (mAlign == 0) //horizontal
        {
            mLayerImage[2].draw(g, -bgX - mLayerX[2], mLayerY[2] - cam->y);
            mLayerImage[2].draw(g, -bgX - mLayerX[2] + 512, mLayerY[2] - cam->y);
            mLayerImage[2].draw(g, -bgX - mLayerX[2] + 1024, mLayerY[2] - cam->y);

            mLayerImage[1].draw(g, -bgX - mLayerX[1], mLayerY[1] - cam->y);
            mLayerImage[1].draw(g, -bgX - mLayerX[1] + 512, mLayerY[1] - cam->y);
            mLayerImage[1].draw(g, -bgX - mLayerX[1] + 1024, mLayerY[1] - cam->y);

            mLayerImage[0].draw(g, -bgX - mLayerX[0], mLayerY[0] - cam->y);
            mLayerImage[0].draw(g, -bgX - mLayerX[0] + 512, mLayerY[0] - cam->y);
            mLayerImage[0].draw(g, -bgX - mLayerX[0] + 1024, mLayerY[0] - cam->y);
        } else //vertical
        {
            mLayerImage[2].draw(g, 0, -bgY);
            mLayerImage[2].draw(g, 0, -bgY + 512);
            mLayerImage[2].draw(g, 0, -bgY + 1024);
        }
    }
    return cam;
}

/**
 * Deallocates memory by destroying the Background
 */
Background::~Background()
{
	mLayerImage[0].~Image();
   	mLayerImage[1].~Image();
   	mLayerImage[2].~Image();
}

File: DragonIsland-Android/src/com/citex/dragonisland/core/drawing/Background.java
package com.citex.dragonisland.core.drawing;

import java.io.IOException;

import javax.microedition.khronos.opengles.GL10;

import com.citex.dragonisland.core.game.Settings;
import com.citex.dragonisland.core.util.FileIO;

/**
	This class draws a scrolling background.
	@author Lawrence Schmid
*/
public class Background {
    
	/** Background image index. */
	private int[] mIndex;
	
    /** Background images. */
    private IBufferedImage mImage[];
    
    /** Background layer x position. */
    private float mX[];
    
    /** Background layer y position. */
    private float mY[];
    
    /** Width of the level. */
    private int mWidth;
    
    /** Height of the level. */
    private int mHeight;
    
    /** Scroll speed. */
    private float mScroll[];
    
    /** Alignment. */
    private int mAlign;
    
    /** Background colour. */
    private Color mColor;
     
    /**
     * Initialises a Background object.
     * @param g Graphics context.
     * @param index Array of image indexes. [0]-near, [1]-middle, [2]-far
     * @param scroll Scroll speed of layers.
     * @param align Background alignment. 0 - horizontal, 1 - vertical
     * @param width Level width.
     * @param height Level Height.
     * @throws IOException 
     */
    public Background(Object g, int index[], float scroll[], int align, int width, int height) throws IOException {

    	// Initialise layer positions.
    	mX = new float[3];
    	mY = new float[3];
    	
    	// Initialise level size.
    	mWidth = width;
    	mHeight = height;
    	
    	// Initialise background variables.
        mAlign = align;
        mScroll = scroll;
        mIndex = index;
        
        // Initialise background layer images.
        mImage = new IBufferedImage[3];
		mImage[0] = FileIO.getImageResource(g, "bgr/near/" + mIndex[0] + ".png");
        mImage[1] = FileIO.getImageResource(g, "bgr/middle/" + mIndex[1] + ".png");
        mImage[2] = FileIO.getImageResource(g, "bgr/far/" + mIndex[2] + ".png"); 

    	// Read the background colour.
    	String[] colour = FileIO.readLine("bgr/far/" + mIndex[2] + ".ini").split(",");
    	mColor = new Color(Integer.parseInt(colour[0]), Integer.parseInt(colour[1]), Integer.parseInt(colour[2]));   	

    	// Initialise the background layer positions.
    	setHeight(height);
    }

    /**
     * Sets the speed of a background layer
     * @param layer The background layer 0-near, 1-middle 2-far
     * @param speed The horizontal scroll speed
     */
    public void setLayerSpeed(int layer, float speed) {
        mScroll[layer] = speed;
    }
    
    /**
     * Paints the background.
     * @param g Graphics context.
     * @param dt Delta time between frame updates.
     * @param cam Camera coordinates.
     * @return Camera coordinates within the screen area.
     */
    public Point draw(Object g, float dt, Point cam) {

    	// Keep the camera within the level area.
    	if (cam.x >= mWidth - Settings.ScreenWidth)
            cam.x = mWidth - Settings.ScreenWidth;
    	else if (cam.x <= 0)
            cam.x = 0;
       	
        if (cam.y >= mHeight - Settings.ScreenHeight)
            cam.y = mHeight - Settings.ScreenHeight;        
        else if (cam.y <= 0)
            cam.y = 0;

        if(!Settings.Background)
        	return cam;
        
        // Set the position of background.
        int bgX = (int)cam.x - ((int)cam.x / 512 * 512);
        int bgY = (int)cam.y - ((int)cam.y / 512 * 512);

        // Scroll background layers.
    	if(!Settings.Paused && Settings.Animation) {
	        for (int i = 0; i < 3; i++) {
	    		if (mX[i] < 512) {
	    			mX[i] += mScroll[i] * dt / 60; 
	    		} else {
	    			mX[i] = 0;
	    		}
	    	}
    	}

        // Draw background.
        if (mAlign == 0) {
        	
        	// Far.
            if(mIndex[2] != 0) {
            	mImage[2].draw(g, -bgX - mX[2], mY[2] - cam.y);
            	mImage[2].draw(g, -bgX - mX[2] + 512, mY[2] - cam.y);
            	mImage[2].draw(g, -bgX - mX[2] + 1024, mY[2] - cam.y);
            }

            // Middle.
            if(mIndex[1] != 0) {
            	mImage[1].draw(g, -bgX - mX[1], mY[1] - cam.y);
            	mImage[1].draw(g, -bgX - mX[1] + 512, mY[1] - cam.y);
            	mImage[1].draw(g, -bgX - mX[1] + 1024, mY[1] - cam.y);
            }

            // Near.
            if(mIndex[0] != 0) {
            	mImage[0].draw(g, -bgX - mX[0], mY[0] - cam.y);
            	mImage[0].draw(g, -bgX - mX[0] + 512, mY[0] - cam.y);
            	mImage[0].draw(g, -bgX - mX[0] + 1024, mY[0] - cam.y);
            }

        } else {
        	// Far.
            mImage[2].draw(g, 0, -bgY);
            mImage[2].draw(g, 0, -bgY + 512);
            mImage[2].draw(g, 0, -bgY + 1024);
        }
        return cam;
    }   

    /**
     * Sets the background colour.
     * @param colour Background colour.
     */
    public void setColor(Color colour) {
        mColor = colour;
    }
    /**
     * Gets the background colour.
     * @return Background colour.
     */
    public Color getColor() {
        return mColor;
    }
    
    /**
     * Sets the alignment of the background.
     * @param align Background alignment. 0 - horizontal, 1 - vertical
     */
    public void setAlign(int align) {
        mAlign = align;
    }

    /**
     * Sets the width of the background
     * @param width The width of the background
     */
    public void setWidth(int width) {
        mWidth = width;
    }
    
    /**
     * Sets the height of the background.
     * @param height Height of the background.
     * @throws IOException 
     * @throws NumberFormatException 
     */
    public void setHeight(int height) throws IOException {
    	    	
        // Read background layer heights.
        int[] imageHeight = new int[3];
        imageHeight[0] = Integer.parseInt(FileIO.readLine("bgr/near/" + mIndex[0] + ".ini").split(",")[3]);
        imageHeight[1] = Integer.parseInt(FileIO.readLine("bgr/middle/" + mIndex[1] + ".ini").split(",")[3]);  		
        imageHeight[2] = Integer.parseInt(FileIO.readLine("bgr/far/" + mIndex[2] + ".ini").split(",")[3]);  		
    	
        // Initialise layer height.
        mY = new float[3];
        mY[0] = height - imageHeight[0];
        mY[1] = height - imageHeight[1];
        mY[2] = height - imageHeight[2];
    }

    /**
     * Destroys the resources.
     * @param gl Graphics context.
     */
    public void destroy(GL10 gl) {
        mImage[0].destroy(gl);
        mImage[1].destroy(gl);
        mImage[2].destroy(gl);
    }
}
